### PR TITLE
Add prettyGrid printer

### LINKED ISSUE #1: Add pretty printing of GameTree
Defining an `Out` instance with http://hackage.haskell.org/package/GenericPretty

### PR DESCRIPTION
**Pull request by momirza**
Fixes #1

### CODE REVIEW

**Comment by momirza** [comment]
To run:

```bash
stack ghci src/Printer.hs 
```

```haskell
*Printer> prettyGameTree 1 (gametree (empty, PlayerO))
Player O
 . | . | . | .
---------------
 . | . | . | .
---------------
 . | . | . | .
---------------
 . | . | . | .
|
+- Player X
|   o | . | . | .
|  ---------------
|   . | . | . | .
|  ---------------
|   . | . | . | .
|  ---------------
|   . | . | . | .
|
+- Player X
|   . | o | . | .
|  ---------------
|   . | . | . | .
|  ---------------
|   . | . | . | .
|  ---------------
|   . | . | . | .
|
+- Player X
|   . | . | o | .
|  ---------------
|   . | . | . | .
|  ---------------
|   . | . | . | .
|  ---------------
|   . | . | . | .
|
+- Player X
|   . | . | . | o
|  ---------------
|   . | . | . | .
|  ---------------
|   . | . | . | .
|  ---------------
|   . | . | . | .
|
+- Player X
|   . | . | . | .
|  ---------------
|   o | . | . | .
|  ---------------
|   . | . | . | .
|  ---------------
|   . | . | . | .
|
| ...
```

File: package.yaml [hunk 1]
```diff
@@ -20,8 +20,10 @@ description:         Please see the README on GitHub at <https://github.com/momi
 dependencies:
 - base >= 4.7 && < 5
 - random
+- prettyprinter
+- containers
 
-default-extensions: MultiWayIf
+default-extensions: MultiWayIf,OverloadedStrings
```
Comment (momirza) [inline package.yaml]: Not sure if I need `OverloadedStrings`

File: src/Printer.hs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import Game hiding (Tree(..))
+import Utils
+
+instance Pretty Cell where
+    pretty X = pretty 'x' 
+    pretty O = pretty 'o'
+    pretty B = pretty '.'
+
+instance Pretty Player where
+    pretty PlayerX = pretty ("Player X" :: String)
```
Comment (momirza) [inline src/Printer.hs]: I presume if I remove `OverloadedStrings` I wouldn't need the type annotation here?

Comment (neongreen) [reply src/Printer.hs]: Yep

File: src/Printer.hs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+prettyTree :: Tree (Grid, Player) -> String
+prettyTree t = drawTree converted
+    where converted = fmap (\(g,p) -> show $ vsep [pretty p, prettyGrid g]) t
+
+prettyGameTree :: Int -> GameTree -> IO () 
+prettyGameTree n = putStrLn . prettyTree . (prune n) . convertTree 
+
+prune :: Int -> Tree a -> Tree a
+prune 0 (Node x _) = Node x []
+prune n (Node x ts) = Node x [prune (n-1) t | t <- ts]
```
Comment (neongreen) [inline src/Printer.hs]: More common version:

```haskell
map (prune (n-1)) ts
```

File: src/Printer.hs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        dashLength = (gridSize - 1) * 4 + 3
+
+convertTree :: GameTree -> Tree (Grid, Player)
+convertTree (GtNode p ts) = Node p $ map convertTree (catMaybes ts)
+
+prettyTree :: Tree (Grid, Player) -> String
+prettyTree t = drawTree converted
+    where converted = fmap (\(g,p) -> show $ vsep [pretty p, prettyGrid g]) t
+
+prettyGameTree :: Int -> GameTree -> IO ()
```
Comment (neongreen) [inline src/Printer.hs:35]: I'd say this function does too many things at once. It's fine for a toy example, but in a bigger project you'd want to separate your code into "library" code with well-defined semantics and "glue"/"app" code where everything goes.

File: src/Printer.hs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    pretty X = pretty 'x' 
+    pretty O = pretty 'o'
+    pretty B = pretty '.'
+
+instance Pretty Player where
+    pretty PlayerX = pretty ("Player X" :: String)
+    pretty PlayerO = pretty ("Player O" :: String)
+
+prettyRow :: Pretty a => [a] -> Doc ann
+prettyRow r = sep $ zipWith (<+>) (emptyDoc : repeat pipe) (map pretty r)
```
Comment (neongreen) [inline src/Printer.hs]: This looks like `hsep $ intersperse pipe (map pretty r)`

File: src/Printer.hs (hunk 4 above)
Comment (momirza) [reply src/Printer.hs:35]: Yep, I agree. 
I just wanted a helper function to easily debug `GameTree`s  so will keep this in for now and probably split it up later if usage becomes problematic.

**Pull request merged**